src/main/java/net/minestom/server/inventory/PlayerInventory.java

```diff
@@ -1,325 +1,187 @@
     }
 
-    @Override
-    public void setChestplate(@NotNull ItemStack itemStack) {
-        safeItemInsert(CHESTPLATE_SLOT, itemStack);
+    private static int getSlotIndex(@NotNull EquipmentSlot slot, int heldSlot) {
+        return switch (slot) {
+            case HELMET, CHESTPLATE, LEGGINGS, BOOTS -> slot.armorSlot();
+            case OFF_HAND -> OFF_HAND_SLOT;
+            case MAIN_HAND -> HOTBAR_START + heldSlot;
+        };
     }
 
-    @Override
-    public @NotNull ItemStack getLeggings() {
-        return getItemStack(LEGGINGS_SLOT);
+    private static @Nullable EquipmentSlot fromSlotIndex(int slot, int heldSlot) {
+        return switch (slot) {
+            case HELMET_SLOT -> EquipmentSlot.HELMET;
+            case CHESTPLATE_SLOT -> EquipmentSlot.CHESTPLATE;
+            case LEGGINGS_SLOT -> EquipmentSlot.LEGGINGS;
+            case BOOTS_SLOT -> EquipmentSlot.BOOTS;
+            case OFF_HAND_SLOT -> EquipmentSlot.OFF_HAND;
+            default -> slot == (HOTBAR_START + heldSlot) ? EquipmentSlot.MAIN_HAND : null;
+        };
     }
 
-    @Override
-    public void setLeggings(@NotNull ItemStack itemStack) {
-        safeItemInsert(LEGGINGS_SLOT, itemStack);
+    private static final int[] EXISTING_ADD_SLOTS = IntStream.concat(
+            IntStream.concat(
+                    IntStream.rangeClosed(36, 44),
+                    IntStream.of(OFF_HAND_SLOT)
+            ),
+            IntStream.rangeClosed(9, 35)
+    ).toArray();
+
+    private static final int[] AIR_ADD_SLOTS = IntStream.concat(
+            IntStream.rangeClosed(36, 44),
+            IntStream.rangeClosed(9, 35)
+    ).toArray();
+
+    private static final int[] TAKE_SLOTS = IntStream.concat(
+            IntStream.rangeClosed(36, 45),
+            IntStream.concat(
+                    IntStream.rangeClosed(9, 35),
+                    IntStream.rangeClosed(0, 8)
+            )
+    ).toArray();
+
+    public PlayerInventory() {
+        super(INVENTORY_SIZE);
     }
 
     @Override
-    public @NotNull ItemStack getBoots() {
-        return getItemStack(BOOTS_SLOT);
+    public byte getWindowId() {
+        return 0;
     }
 
     @Override
-    public void setBoots(@NotNull ItemStack itemStack) {
-        safeItemInsert(BOOTS_SLOT, itemStack);
-    }
+    public void updateSlot(int slot, @NotNull ItemStack itemStack) {
+        super.updateSlot(slot, itemStack);
 
-    /**
-     * Refreshes the player inventory by sending a {@link WindowItemsPacket} containing all.
-     * the inventory items
-     */
-    @Override
-    public void update() {
-        this.player.sendPacket(createWindowItemsPacket());
-    }
+        for (var player : getViewers()) {
+            var equipmentSlot = fromSlotIndex(slot, player.getHeldSlot());
+            if (equipmentSlot == null) continue;
 
-    /**
-     * Gets the item in player cursor.
-     *
-     * @return the cursor item
-     */
-    public @NotNull ItemStack getCursorItem() {
```

Comment:
I think supporting multiple players opening the same inventory is important though, not necessarily from a practical perspective (where people would be constantly using this feature) but to solidify what we mean as an inventory, since now a lot of features from Inventory would be essentially unused by PlayerInventory.

Reply:
Im not sure which features exactly, but in that case I believe they should just be in ContainerINventory if they are unused in PlayerInventory

Reply:
I am probably mistaken, it seems fine.